topics/go/algorithms/data/hash/hash.go

```diff
@@ -0,0 +1,107 @@
+// Package hash implements a hash table
+package hash
+
+import (
+	"fmt"
+	"hash/maphash"
+)
+
+const (
+	numBuckets = 256 // Number of buckets in hash
+)
+
+// entry in the hash table
+type entry struct {
+	key   string
+	value int
+}
+
+// Hash is a simple Hash table
+type Hash struct {
+	buckets [][]*entry
+	h       maphash.Hash
+}
+
+// NewHash returns a new hash table
+func NewHash() *Hash {
+	return &Hash{
+		buckets: make([][]*entry, numBuckets),
+	}
+}
+
+// Set sets the key to value in the hash
+func (h *Hash) Set(key string, value int) {
+	b := h.hashKey(key)
+	// Look for existing key
+	for _, e := range h.buckets[b] {
+		if e.key == key {
+			e.value = value
+			return
+		}
+	}
+
+	h.buckets[b] = append(h.buckets[b], &entry{key, value})
```

Comment:
Now I defintely want to use value semantics since we own the value. Keep the data closer together.

Reply:
See above comment

Reply:
We can use value semantic mutation and just replace the value.

Reply:
OK, changed with a TODO for a good comment on why we create a new entry